### ISSUE TITLE
Is SplitHorizon functional? External access failure

### ISSUE THREAD

**Issue by puertal** [mongodb/mongodb-kubernetes-operator#483]
**What did you do to encounter the bug?**
Deploy the next MongoDBCommunity resource:
```
apiVersion: mongodbcommunity.mongodb.com/v1
kind: MongoDBCommunity
metadata:
  name: ds-mongodb
  namespace: test-ds
spec:
  members: 3
  replicaSetHorizons:
  - test-horizon-1: node1:31110
    test-horizon-2: ds-mongodb-svc-0.test-ds.svc.cluster.local:27017
  - test-horizon-1: node2:31111
    test-horizon-2: ds-mongodb-svc-1.test-ds.svc.cluster.local:27017
  - test-horizon-1: node3:31112
    test-horizon-2: ds-mongodb-svc-2.test-ds.svc.cluster.local:27017
  security:
    authentication:
      modes:
      - SCRAM
      tls:
        enabled: true
  type: ReplicaSet
  users:
  - db: admin
    name: user
    passwordSecretRef:
      name: user-ds-mongodb
    roles:
    - db: admin
      name: clusterAdmin
    - db: admin
      name: userAdminAnyDatabase
    - db: ds
      name: readWrite
    scramCredentialsSecretName: ds-mongodb
  version: 4.2.6
```

**What did you expect?**
The creation of a 3 nodes MongoDB cluster accessible from outside with this connection:
`mongo --host ds-mongodb/node1:31110,node2:31111,node3:31112
`

**What happened instead?**
MongoDB still renames the hosts to the internal ones and the connection fails:
```
MongoDB shell version v3.6.8
connecting to: mongodb://node1:31110,node2:31111,node3:31112/?replicaSet=ds-mongodb
2021-05-12T10:06:20.826+0200 I NETWORK  [thread1] Starting new replica set monitor for ds-mongodb/node1:31110,node2:31111,node3:31112
2021-05-12T10:06:20.836+0200 I NETWORK  [thread1] Successfully connected to node2:31111 (1 connections now open to node2:31111 with a 5 second timeout)
2021-05-12T10:06:20.837+0200 I NETWORK  [ReplicaSetMonitor-TaskExecutor-0] Successfully connected to node3:31112 (1 connections now open to node3:31112 with a 5 second timeout)
2021-05-12T10:06:20.840+0200 I NETWORK  [thread1] changing hosts to ds-mongodb/ds-mongodb-0.ds-mongodb-svc.test-ds.svc.cluster.local:27017,ds-mongodb-1.ds-mongodb-svc.test-ds.svc.cluster.local:27017,ds-mongodb-2.ds-mongodb-svc.test-ds.svc.cluster.local:27017 from ds-mongodb/node1:31110,node2:31112,node3:31111
2021-05-12T10:06:20.849+0200 I NETWORK  [thread1] getaddrinfo("ds-mongodb-0.ds-mongodb-svc.test-ds.svc.cluster.local") failed: Name or service not known
2021-05-12T10:06:20.853+0200 I NETWORK  [ReplicaSetMonitor-TaskExecutor-0] getaddrinfo("ds-mongodb-2.ds-mongodb-svc.test-ds.svc.cluster.local") failed: Name or service not known
2021-05-12T10:06:20.857+0200 I NETWORK  [thread1] getaddrinfo("ds-mongodb-1.ds-mongodb-svc.test-ds.svc.cluster.local") failed: Name or service not known
2021-05-12T10:06:21.360+0200 I NETWORK  [thread1] getaddrinfo("ds-mongodb-1.ds-mongodb-svc.test-ds.svc.cluster.local") failed: Name or service not known
2021-05-12T10:06:21.363+0200 I NETWORK  [thread1] getaddrinfo("ds-mongodb-2.ds-mongodb-svc.test-ds.svc.cluster.local") failed: Name or service not known
2021-05-12T10:06:21.366+0200 I NETWORK  [thread1] getaddrinfo("ds-mongodb-0.ds-mongodb-svc.test-ds.svc.cluster.local") failed: Name or service not known
2021-05-12T10:06:21.366+0200 W NETWORK  [thread1] Unable to reach primary for set ds-mongodb
```

**Screenshots**

**Operator Information**
 - Operator Version 0.6.0
 - MongoDB Image used registry.hub.docker.com/library/mongo:4.2.6

**Kubernetes Cluster Information**
 - Distribution Rancher installation on VMs.
 - Version v1.19.6
 - Image Registry location (quay, or an internal registry) DockerHub

**Additional context**

If possible, please include:
This node logs imply that the Split Horizon has been applied:
```
[2021-05-12T08:02:35.048+0000] [.info] [src/action/dbcmd.go:replSetInitiateWithConf:103] <ds-mongodb-0> [08:02:35.048] Running ReplSetInitiate : map[_id:ds-mongodb members:[map[_id:0 arbiterOnly:false buildIndexes:true hidden:false horizons:map[test-horizon-1:node1:31110 test-horizon-2:ds-mongodb-svc-0.test-ds.svc.cluster.local:27017] host:ds-mongodb-0.ds-mongodb-svc.test-ds.svc.cluster.local:27017 priority:1 slaveDelay:0 tags:map[] votes:1] map[_id:1 arbiterOnly:false buildIndexes:true hidden:false horizons:map[test-horizon-1:node2:31111 test-horizon-2:ds-mongodb-svc-1.test-ds.svc.cluster.local:27017] host:ds-mongodb-1.ds-mongodb-svc.test-ds.svc.cluster.local:27017 priority:1 slaveDelay:0 tags:map[] votes:1] map[_id:2 arbiterOnly:false buildIndexes:true hidden:false horizons:map[test-horizon-1:node3:31112 test-horizon-2:ds-mongodb-svc-2.test-ds.svc.cluster.local:27017] host:ds-mongodb-2.ds-mongodb-svc.test-ds.svc.cluster.local:27017 priority:1 slaveDelay:0 tags:map[] votes:1]] protocolVersion:1]
[2021-05-12T08:02:35.400+0000] [.info] [src/action/dbcmd.go:replSetInitiateWithConf:115] <ds-mongodb-0> [08:02:35.400] <DB_WRITE> Initiated replica set.  Cmd: [{replSetInitiate map[_id:ds-mongodb members:[map[_id:0 arbiterOnly:false buildIndexes:true hidden:false horizons:map[test-horizon-1:node1:31110 test-horizon-2:ds-mongodb-svc-0.test-ds.svc.cluster.local:27017] host:ds-mongodb-0.ds-mongodb-svc.test-ds.svc.cluster.local:27017 priority:1 slaveDelay:0 tags:map[] votes:1] map[_id:1 arbiterOnly:false buildIndexes:true hidden:false horizons:map[test-horizon-1:node2:31111 test-horizon-2:ds-mongodb-svc-1.test-ds.svc.cluster.local:27017] host:ds-mongodb-1.ds-mongodb-svc.test-ds.svc.cluster.local:27017 priority:1 slaveDelay:0 tags:map[] votes:1] map[_id:2 arbiterOnly:false buildIndexes:true hidden:false horizons:map[test-horizon-1:node3:31112 test-horizon-2:ds-mongodb-svc-2.test-ds.svc.cluster.local:27017] host:ds-mongodb-2.ds-mongodb-svc.test-ds.svc.cluster.local:27017 priority:1 slaveDelay:0 tags:map[] votes:1]] protocolVersion:1]}]
```

**Comment by spike008t** [comment]
I have the same issue

**Comment by dgdevops** [comment]
Same issue here

**Comment by vasRK** [comment]
not sure about SplitHorizon i access mongodb outside cluster using a loadbalancer

`
apiVersion: v1
kind: Service
metadata:
annotations:
service.beta.kubernetes.io/azure-dns-label-name: example-mongodb-prod
name: example-mongodb-prod-lb
spec:
loadBalancerIP: 111.111.111.111
type: LoadBalancer
selector:
statefulset.kubernetes.io/pod-name: example-mongodb-0
ports:
- protocol: TCP
port: 27017
targetPort: 27017

`

I did it with a cluster loadbalancer, i use azure kubernetes service, im able to access mongodb using assigned dns name or ip both from mongo shell & mongo compass

**Comment by puertal** [comment]
Dear @vasRK , you are configuring one POD inside the StatefulSet.
If this POD fails or is not the Master node I think you won't be able to connect with the Mongo cluster.
You do not have HA.
That is why we need to use SplitHorizon to access the cluster from outside the K8s environment (DBAdmins).
Regards.

**Comment by vasRK** [comment]
@puertal Good to know,  I use this only to manually verify data or query, all my apps are on same cluster for now. may be i need SplitHorizon when i want to connect to replicaset from other clusters.

thanks

**Comment by dgdevops** [comment]
Is there any ETA for fixing this?

**Comment by dgdevops** [comment]
Is there any update here @bznein?

**Comment by puertal** [comment]
Somebody there?

**Comment by repocho** [comment]
I have same issue 😢 
Will it be fixed in next release ??

Thank you !!

**Comment by priyolahiri** [comment]
There's work in progress to redesign the feature for community operator. The specs will change, and we intend to add the ability for the services to be created for external access as well to reduce the manual steps required.

**Comment by dgdevops** [comment]
Knowing the release date would definitely be a great help considering the fact that this issue is blocking our developments.

**Comment by dgdevops** [comment]
@priyolahiri, by any chance is there an ETA for the redesigned feature to be released?
Thank you

**Comment by priyolahiri** [comment]
Sorry, there is no exact ETA at the moment, but it is WIP.

**Comment by dgdevops** [comment]
Thank you for the reply @priyolahiri , I upgraded the Operator to the latest version but I am still experiencing the issue described in this ticket. I can access each member separately from outside of the Kubernetes cluster but I cannot reach the replicaset:

```
/ # mongo "mongodb://public-ip-0:27017,public-ip-1:27017,public-ip-2:27017/?replicaSet=rs-name" --ssl --sslAllowInvalidCertificates
MongoDB shell version v4.0.5
connecting to: mongodb://public-ip-0:27017,public-ip-1:27017,public-ip-2:27017/?gssapiServiceName=mongodb&replicaSet=rs-name
2021-07-06T12:17:47.715+0000 I NETWORK  [js] Starting new replica set monitor for rs-name/public-ip-0:27017,public-ip-1:27017,public-ip-2:27017
2021-07-06T12:17:47.729+0000 W NETWORK  [ReplicaSetMonitor-TaskExecutor] SSL peer certificate validation failed: self signed certificate
2021-07-06T12:17:47.730+0000 W NETWORK  [js] SSL peer certificate validation failed: self signed certificate
2021-07-06T12:17:47.733+0000 I NETWORK  [ReplicaSetMonitor-TaskExecutor] Successfully connected to public-ip-0:27017 (1 connections now open to public-ip-0:27017 with a 5 second timeout)
2021-07-06T12:17:47.734+0000 I NETWORK  [js] Successfully connected to public-ip-1:27017 (1 connections now open to public-ip-1:27017 with a 5 second timeout)
2021-07-06T12:17:47.812+0000 I NETWORK  [js] changing hosts to rs-name/rs-name-0.rs-name-svc.mongodb.svc.cluster.local:27017,rs-name/rs-name-1.rs-name-svc.mongodb.svc.cluster.local:27017,rs-name/rs-name-2.rs-name-svc.mongodb.svc.cluster.local:27017:27017 from rs-name/public-ip-2:27017,public-ip-1:27017,public-ip-0:27017
2021-07-06T12:17:48.518+0000 W NETWORK  [js] Unable to reach primary for set rs-name
```

Configuration:
```
  replicaSetHorizons:
  - horizon: rs-name-0.public-domain:27017
  - horizon: rs-name-1.public-domain:27017
  - horizon: rs-name-2.public-domain:27017
```

Is there a way to get rid of this error?

**Comment by priyolahiri** [comment]
You would want to create nodePort services for the replicaset for external access. Use alternate ports, for example 31181, 31182, 31183

Your split horizon config would look like 

`
replicaSetHorizons:
  - horizon: node1.public-ip.com:31181
  - horizon: node2.public-ip.com:31182
  - horizon: node3.public-ip.com:31183
`
Make sure the certs have the required SANs. 

Your mongo connection would then look like:
`
mongo "mongodb://node1.public-ip.com:31181,node2.public-ip.com:31182,node3.public-ip.com:31183/?replicaSet=rs-name" --ssl --sslAllowInvalidCertificates
`

**Comment by dgdevops** [comment]
Hello @priyolahiri ,
Thank you for your response.

The Operator & ReplicaSets run on GKE clusters where all our services are exposed with "LoadBalancer" service type (either Internal/External). Exposing services with "NodePort" on managed Kubernetes clusters is possible but no external IP is associated with the services exposed this way hence the internal/external LoadBalancer service type is in use. With the "LoadBalancer" service type I get the same error that @puertal described while he was using "NodePort" service.

**Comment by priyolahiri** [comment]
New doc added: ```docs/external_access.md``` and samples in ```config/samples/external_access```

**Comment by dgdevops** [comment]
Hello @priyolahiri,
Thank you for the documentation.
I have managed to set up the feature successfully using LoadBalancer service type.

Few observations regarding the documentation:
1) `docs/external_access.md` - `--tlsCAfile` should be `--tlsCAFile` (as `file` with lowercase is not an accepted flag)
2) When configuring the certificate the commonName field needs to contain a domain with .local TLD so the ticket must be self-signed (not trusted) as a result `SSL peer certificate validation failed: unable to verify the first certificate` error appears. In order to get rid of the error I need to use `--tlsAllowInvalidCertificates` flag to make it work. Is this expected?
Note: I see the error when I use MongoDB shell version `v4.2.10` , when I change to an older version such as `v4.0.5` I don't see the error.

**Comment by repocho** [comment]
Hello,

is this still not working or was fixed in a release ?

Thanks.w

**Comment by bznein** [comment]
Hi @repocho,

please see https://github.com/mongodb/mongodb-kubernetes-operator/issues/483#issuecomment-878319275

Kind regards,
Nikolas

**Comment by jamesbroadhead** [comment]
Documented in #483. Please file a ticket if following those docs is not successful

**Comment by maxisam** [comment]
@dgdevops Could you share how you config LoadBalancer service type for external access? Thanks!

**Comment by dgdevops** [comment]
Hey @maxisam,
I am using internal LoadBalancer service with `cloud.google.com/load-balancer-type=Internal` annotation on GKE that works well with the replicaSetHorizons setup that I have configured this way:
```
  replicaSetHorizons:
  - horizon: ${member0_domain_name}:${member0_svc_port}
  - horizon: ${member1_domain_name}:${member1_svc_port}
  - horizon: ${member2_domain_name}:${member2_svc_port}
```
The internal LoadBalancer services (3 in total, one for each MongoDB member) are configured for the FQDN specified in the replicaSetHorizons configuration (member0/1/2_domain_name) with external DNS and have static ports specified that are also referenced in the config above.

To sum up:
- 3 LoadBalancers (in my case internal) one for each node (with custom FQDN & port)
- replicaSetHorizons configuration (3 entries, one for each node) pointing to the LoadBalancer FQDNs & custom ports

**Comment by saimantr** [comment]
Hello @dgdevops , did you hear about any reply/ workaround ? I am in the same state as you are. I have a replicaset with 3 members and exposing each of the member with a loadbalancer service.  I can access each member separately from outside of the Kubernetes cluster but I cannot reach the replicaset.

Thanks,
Sai
